python/python/bystro/supervised_ppca/gf_adversarial_pt.py

```diff
@@ -0,0 +1,318 @@
+            y_batch = y_[idx]
+
+            sigma = softplus(sigmal_)
+            WWT = torch.matmul(torch.transpose(W_, 0, 1), W_)
+            Sigma = WWT + sigma * eye
+
+            like_prior = _prior(trainable_variables)
+
+            m = MultivariateNormal(zerosp, Sigma)
+            like_gen = torch.mean(m.log_prob(X_batch))
+
+            P_x, Cov = _get_projection_matrix(W_, sigma)
+            mean_z = torch.matmul(X_batch, torch.transpose(P_x, 0, 1))
+            eps = torch.rand_like(mean_z)
+            C1_2 = torch.linalg.cholesky(Cov)
+            z_samples = mean_z + torch.matmul(eps, C1_2)
+
+            y_hat = torch.squeeze(discriminator(z_samples))
+
+            if task == "regression":
+                loss_y = supervision_loss(y_hat, y_batch)
+            else:
+                loss_y = supervision_loss(sigm(y_hat), y_batch)
+
+            optimizer_d.zero_grad()
+            loss_y.backward(retain_graph=True)
+            optimizer_d.step()
+
+            y_hat2 = torch.squeeze(discriminator(z_samples))
+            if task == "regression":
+                loss_y2 = supervision_loss(y_hat2, y_batch)
+            else:
+                loss_y2 = supervision_loss(sigm(y_hat2), y_batch)
+
+            WTW = torch.matmul(W_, torch.transpose(W_, 0, 1))
+            off_diag = WTW - torch.diag(torch.diag(WTW))
+            loss_i = torch.linalg.matrix_norm(off_diag)
+
+            posterior = like_gen + 1 / N * like_prior
+            loss_gen = -1 * posterior - self.mu * loss_y2 + self.gamma * loss_i
+
+            optimizer_g.zero_grad()
+            loss_gen.backward()
+            optimizer_g.step()
+
+            self._save_losses(i, like_gen, like_prior, posterior)
+            self.losses_supervision[i] = loss_y.detach().numpy()
+
+        self._store_instance_variables(trainable_variables)
+
+        return self
+
+    def _store_instance_variables(self, trainable_variables: list[Tensor]):
+        """
+        Saves the learned variables
+
+        Parameters
+        ----------
+        trainable_variables : list
+            List of saved variables of type Tensor
+
+        Sets
+        ----
+        W_ : NDArray,(n_components,p)
+            The loadings
+
+        sigma2_ : float
+            The isotropic variance
+
+        B_ : float
+            The intercept for the predictive model
+        """
+        self.W_ = trainable_variables[0].detach().numpy()
+        self.sigma2_ = nn.Softplus()(trainable_variables[1]).detach().numpy()
+
+    def _initialize_variables(self, X: NDArray):
+        """
+        Initializes the variables of the model. Right now fits a PCA model
+        in sklearn, uses the loadings and sets sigma^2 to be unexplained
+        variance.
+
+        Parameters
+        ----------
+        X : NDArray,(n_samples,p)
+            The data
+
+        Returns
+        -------
+        W_ : torch.tensor,shape=(n_components,p)
+            The loadings of our latent factor model
```

Comment:
Do we want to keep documentation specific to factor models?

Reply:
What do you mean?